mitmproxy/proxy/protocol2/websocket.py

```diff
@@ -16,14 +18,98 @@ def __init__(self, context: ClientServerContext, handshake_flow: http.HTTPFlow):
+
+        self.server_conn = wsconn.WSConnection(wsconn.CLIENT, wsconn.ConnectionState.OPEN,
+                                               extensions=[extension] if extension else None)
 
     @expect(events.Start)
     def start(self, _) -> commands.TCommandGenerator:
-        yield from ()
+        yield commands.Hook("websocket_start", self.flow)
         self._handle_event = self.relay_messages
 
+    _handle_event = start
+
     @expect(events.DataReceived, events.ConnectionClosed)
     def relay_messages(self, event: events.Event) -> commands.TCommandGenerator:
-        raise NotImplementedError()
+        if isinstance(event, events.DataReceived):
+            from_client = event.connection == self.context.client
+            if from_client:
+                source = self.client_conn
+                other = self.server_conn
+                fb = self.client_frame_buffer
+                send_to = self.context.server
+            else:
+                source = self.server_conn
+                other = self.client_conn
+                fb = self.server_frame_buffer
+                send_to = self.context.client
 
-    _handle_event = start
+            source.receive_bytes(event.data)
+
+            for ws_event in source.events():
+                if isinstance(ws_event, wsevents.DataReceived):
+                    fb.append(ws_event.data)
+                    if ws_event.message_finished:
+                        if isinstance(ws_event, wsevents.BytesReceived):
+                            payload = b"".join(fb)
+                        else:
+                            payload = "".join(fb)
+                        fb.clear()
+                        websocket_message = websocket.WebSocketMessage(0x1 if isinstance(ws_event, wsevents.TextReceived) else 0x2,
+                                                                       from_client, payload)
+                        self.flow.messages.append(websocket_message)
+                        yield commands.Hook("websocket_message", self.flow)
+
+                    other.send_data(ws_event.data, ws_event.message_finished)
+                    yield commands.SendData(send_to, other.bytes_to_send())
+                elif isinstance(ws_event, (wsevents.PingReceived, wsevents.PongReceived)):
+                    yield commands.Log(
+                        "info",
+                        "Websocket {} received {}".format(
+                            "PING" if isinstance(ws_event, wsevents.PingReceived) else "PONG",
+                            ws_event.payload.decode()
+                        )
+                    )
+                    if isinstance(ws_event, wsevents.PingReceived):
+                        other.ping(ws_event.payload)
+                        yield commands.SendData(send_to, other.bytes_to_send())
+                        source.bytes_to_send()  # We don't want send PONG unless we receive it
```

Comment:
> Let's trigger a ping to other

I would recommend to just reply with a pong to source, I don't think there's a need to forward that to the server. After all, that'd be a communication channel through mitmproxy that is not logged.

> So, what should we do on a PongReceived event?


The RFC says that "A Pong frame MAY be sent unsolicited.", so let's just do nothing.